ssas/service/blacklist/blacklist.go

```diff
@@ -0,0 +1,325 @@
+//		until the cached blacklist is refreshed from the database.
+func (b *blacklist) IsTokenBlacklisted(tokenID string) bool {
+	b.RLock()
+	defer b.RUnlock()
+
+	bEvent := ssas.Event{Op: "TokenVerification", TrackingID: b.ID, TokenID: tokenID}
+	if _, found := b.tc.Get(tokenID); found {
+		ssas.BlacklistedTokenPresented(bEvent)
+		return true
+	}
+	return false
+}
+
+func (b *blacklist) BlacklistGroup(field GroupField, value string) error {
+	entryDate := time.Now()
+	expirationDate := entryDate.Add(b.gcLifetime)
+	entry, err := createGroupEntry(value, entryDate, expirationDate, field)
+	if err != nil {
+		return fmt.Errorf(fmt.Sprintf("unable to create group blacklist entry value %s: %s", value, err.Error()))
+	}
+
+	// Add to cache only after token is blacklisted in database
+	ssas.GroupBlacklisted(ssas.Event{Op: "GroupBlacklist", TrackingID: strconv.FormatUint(uint64(entry.ID), 10)})
+
+	b.gc.Set(uuid.New(), groupEntry{field, value}, b.gcLifetime)
+
+	return nil
+}
+
+func (b *blacklist) IsGroupBlacklisted(group ssas.Group) bool {
+	var cacheItems map[string]cache.Item
+
+	b.RLock()
+	cacheItems = b.gc.Items()
+	b.RUnlock()
+
+	var xDataObj map[string]interface{}
+	if err := json.Unmarshal([]byte(group.XData), &xDataObj); err != nil {
+		ssas.OperationFailed(ssas.Event{Op: "XDataUnmarshal", Help: err.Error()})
+	}
+
+	for _, item := range cacheItems {
+		entry, ok := item.Object.(groupEntry)
+		if !ok {
+			continue
+		}
+
+		switch entry.field {
+		case GroupFieldXData:
+			// Nothing to compare against
+			if xDataObj == nil {
+				continue
+			}
+			var expressionObj map[string]interface{}
+			if err := json.Unmarshal([]byte(entry.expression), &expressionObj); err != nil {
+				continue
+			}
+			if checkXData(xDataObj, expressionObj) {
+				ssas.BlacklistedGroupPresented(ssas.Event{Op: "GroupVerification", TrackingID: group.GroupID})
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (b *blacklist) startCacheRefreshTicker(refreshFreq time.Duration, refreshFunc func() error) (ticker *time.Ticker, done chan struct{}) {
+	event := ssas.Event{Op: "CacheRefreshTicker", TrackingID: b.ID}
+	ssas.ServiceStarted(event)
+
+	ticker = time.NewTicker(refreshFreq)
+	done = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				if err := refreshFunc(); err != nil {
+					ssas.CacheSyncFailure(ssas.Event{Op: "BlacklistLoadFromDatabase", TrackingID: b.ID, Help: err.Error()})
+				}
+			}
+		}
+	}()
+
+	return
+}
+
+func (b *blacklist) loadTokensFromDatabase() error {
+	var (
+		entries []TokenEntry
+		err     error
+	)
+
+	if entries, err = getUnexpiredTokenEntries(); err != nil {
+		return err
+	}
+
+	// Need to acquire a lock since we're clearing the entire cache.
+	// Any reads in-between us re-hydrating the cache is invalid (false negatives)
+	b.Lock()
+	defer b.Unlock()
+
+	b.tc.Flush()
```

Comment:
The [old implementation ](https://github.com/CMSgov/bcda-ssas-app/pull/56/files#diff-020d650a2ce74e27bb6e3de0c7d301c9c192a5c033981502f474e570a695e79eL106) did not use a mutex to restrict access to the cache while it was being rebuilt. This may result in false negatives being reported (token is blacklisted, but we **do not** find it in the cache).

If a caller checks to see if the token is blacklisted after we've flushed (cleared) the cache and before we've finished rebuilding the cache, we would incorrectly say that the token **is not** blacklisted.

By acquiring a write lock, we prevent callers from reading the cache until it's completely rebuilt.